test/gtest/ir-traversal-test.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2024-present Intel Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "ir/ir-traversal.h"
+
+#include <gtest/gtest.h>
+
+#include "ir/ir.h"
+
+namespace P4::Test {
+
+struct TraversalTest : public ::testing::Test {};
+
+TEST_F(TraversalTest, SimpleIRApply) {
+    const auto *v = new IR::Constant(42);
+    const auto *m = IR::Traversal::apply(v, &IR::Constant::value, IR::Traversal::Assign(4));
+    EXPECT_NE(v, m);
+    EXPECT_EQ(v->value, 42);
+    EXPECT_EQ(m->value, 4);
+
+    const auto *t = IR::Type_Bits::get(4);
+    const auto *m2 = IR::Traversal::apply(v, &IR::Constant::type, IR::Traversal::Assign(t));
+    EXPECT_NE(v, m2);
+    EXPECT_EQ(v->value, 42);
+    EXPECT_EQ(m2->value, 42);
+    EXPECT_EQ(m2->type, t);
+}
+
+TEST_F(TraversalTest, SimpleIRModify) {
+    auto *v = new IR::Constant(42);
+    auto *m = IR::Traversal::modify(v, &IR::Constant::value, IR::Traversal::Assign(4));
+    EXPECT_EQ(v, m);
+    EXPECT_EQ(v->value, 4);
+
+    const auto *t = IR::Type_Bits::get(4);
+    const auto *m2 = IR::Traversal::modify(v, &IR::Constant::type, IR::Traversal::Assign(t));
+    EXPECT_EQ(v, m2);
+    EXPECT_EQ(v->value, 4);
+    EXPECT_EQ(v->type, t);
+}
+
+TEST_F(TraversalTest, ComplexIRModify) {
+    const auto *path = new IR::Path("foo");
+    const auto *pe = new IR::PathExpression(path);
+    const auto *add = new IR::Add(pe, new IR::Constant(42));
+    auto *asgn = new IR::AssignmentStatement(pe, add);
+    IR::StatOrDecl *stmt = asgn;
+
+    modify(stmt, RTTI::to<IR::AssignmentStatement>, &IR::AssignmentStatement::right,
+           RTTI::to<IR::Operation_Binary>, &IR::Operation_Binary::left,
+           RTTI::to<IR::PathExpression>, &IR::PathExpression::path, &IR::Path::name, &IR::ID::name,
+           IR::Traversal::Assign(P4::cstring("bar")));
```

Comment:
This seems very ugly/hard to use, and seems like it could benefit from somthing like the IR::pattern stuff in ir/pattern.h.  Not sure how that could work, but if you could write this more like

    modify(stmt, Assignment(AnyExpression, placeholder + AnyExpression), something to set placeholder....

it would be much cleaner.

Reply:
I'd say it is not nice, but it is nicer than the alternative:
- you could write the clones and assignments manually, which is very ugly, much uglier than this, and it hard to see if it is correct;
- you could visit the argument and then match on its context, which would be more expensive and if you need to modify multiple sub-expressions, you would end up with very spread code which would also run the checks on context many times.


Of course, I would like it very much if I could just pass the "field chain" (e.g. `obj->left->type`) to a function like that, but that is not possible. The member pointers are clunky.

As for pattern-like style -- the disadvantage of `Pattern` is that it is a separate partial mirror image of the IR. Also, I think that the pattern could get quite complicated if we tried to allow casts in them. If we wanted to have something like patterns, it would make sense to study what was proposed for the C++ pattern matching (because they tried to be quite general) and to try to create some way to generate the patters in IR generator. But that would be a major undertaking, while this one was quite quick to do (and unlike `patter.h` it is documented).

Reply:
Also note, that the pattern syntax gets poorly readable quite fast -- already in your example, one must guess that the two parameters of `Assignment` are `left` and `right` and that `type`, `srcInfo` and `annotations` are not shown in the pattern -- but what if there is need to modify these? This quill get quite tricky once more nested patterns come into play (also because one has to somehow represent them in C++ which is not meant for this).

Reply:
Yes, unfortunately C++ does not have any decent of doing patterns that doesn't result in excessive runtime overhead (no good macros or ways to ensure stuff happens at compile time).  Its always worth thinking about trying to come up with novel ways to get around that, but this seems like it at least improves some things already.